distribution/pull_v2.go

```diff
@@ -98,91 +93,114 @@ func (p *v2Puller) pullV2Repository(ref reference.Named) (err error) {
+	logrus.Debugf("pulling blob %q", ld.digest)
 
-	blobs := p.repo.Blobs(context.Background())
+	blobs := ld.repo.Blobs(ctx)
 
-	desc, err := blobs.Stat(context.Background(), di.digest)
+	desc, err := blobs.Stat(ctx, ld.digest)
 	if err != nil {
 		logrus.Debugf("Error statting layer: %v", err)
-		di.err <- err
-		return
+		if err == distribution.ErrBlobUnknown {
+			return nil, 0, xfer.DoNotRetry{err}
+		}
+		return nil, 0, err
 	}
-	di.size = desc.Size
 
-	layerDownload, err := blobs.Open(context.Background(), di.digest)
+	layerDownload, err := blobs.Open(ctx, ld.digest)
 	if err != nil {
 		logrus.Debugf("Error fetching layer: %v", err)
-		di.err <- err
-		return
+		return nil, 0, err
 	}
 	defer layerDownload.Close()
 
-	verifier, err := digest.NewDigestVerifier(di.digest)
+	verifier, err := digest.NewDigestVerifier(ld.digest)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	tmpFile, err := ioutil.TempFile("", "GetImageBlob")
 	if err != nil {
-		di.err <- err
-		return
+		return nil, 0, err
 	}
 
-	digestStr := di.digest.String()
+	reader := xfer.NewProgressReader(ioutil.NopCloser(io.TeeReader(layerDownload, verifier)), progressChan, desc.Size, ld.ID(), "Downloading")
 
-	reader := progressreader.New(progressreader.Config{
-		In:        ioutil.NopCloser(io.TeeReader(layerDownload, verifier)),
-		Out:       di.broadcaster,
-		Formatter: p.sf,
-		Size:      di.size,
-		NewLines:  false,
-		ID:        stringid.TruncateID(digestStr),
-		Action:    "Downloading",
-	})
-	io.Copy(di.tmpFile, reader)
+	transferDone := make(chan struct{})
+	go func() {
+		select {
+		case <-ctx.Done():
+			layerDownload.Close()
+		case <-transferDone:
+			return
+		}
+	}()
+	io.Copy(tmpFile, reader)
+	close(transferDone)
 
-	di.broadcaster.Write(p.sf.FormatProgress(stringid.TruncateID(digestStr), "Verifying Checksum", nil))
+	select {
+	case <-ctx.Done():
+		return nil, 0, xfer.DoNotRetry{ctx.Err()}
+	default:
+	}
+
+	progressChan <- xfer.Progress{ID: ld.ID(), Action: "Verifying Checksum"}
 
 	if !verifier.Verified() {
-		err = fmt.Errorf("filesystem layer verification failed for digest %s", di.digest)
+		err = fmt.Errorf("filesystem layer verification failed for digest %s", ld.digest)
 		logrus.Error(err)
-		di.err <- err
-		return
+		tmpFile.Close()
+		if err := os.RemoveAll(tmpFile.Name()); err != nil {
+			logrus.Errorf("Failed to remove temp file: %s", tmpFile.Name())
+		}
+
+		return nil, 0, err
 	}
 
-	di.broadcaster.Write(p.sf.FormatProgress(stringid.TruncateID(digestStr), "Download complete", nil))
+	progressChan <- xfer.Progress{ID: ld.ID(), Action: "Download complete"}
 
-	logrus.Debugf("Downloaded %s to tempfile %s", digestStr, di.tmpFile.Name())
-	di.layer = layerDownload
+	logrus.Debugf("Downloaded %s to tempfile %s", ld.ID(), tmpFile.Name())
 
-	di.err <- nil
+	tmpFile.Seek(0, 0)
+	return ioutils.NewReadCloserWrapper(tmpFile, tmpFileCloser(tmpFile)), desc.Size, nil
```

Comment:
I implemented a cancellable progress reader in https://github.com/aaronlehmann/docker/commit/1d9fb7ea14f94877f1328d8133e8923e4bf2110b. Does this look good?

In this same commit, I changed `layer.Layer`'s `TarStream` to return an `io.ReadCloser` so that uploads could be safely interrupted. I'm thinking about taking this part of the commit and making a PR against master, because the current code on master can leak if an upload or export fails partway through. @tonistiigi, what do you think?